src/Symfony/Bundle/DebugBundle/VarDumper/DelayedDumper.php

```diff
@@ -0,0 +1,121 @@
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bundle\DebugBundle\VarDumper;
+
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\HttpKernel\Event\FilterResponseEvent;
+use Symfony\Component\HttpKernel\KernelEvents;
+use Symfony\Component\VarDumper\Cloner\Data;
+use Symfony\Component\VarDumper\Dumper\CliDumper;
+use Symfony\Component\VarDumper\Dumper\DataDumperInterface;
+use Symfony\Component\VarDumper\Dumper\HtmlDumper;
+
+/**
+ * Used as fallback in the ServerDumper, it delays output to the end of the process for web requests.
+ * It replicates the DumpDataCollector behavior to ensure dumps are written to the output whenever the dump server isn't up.
+ * In cli, output is not delayed and the provided CliDumper instance is used instead.
+ *
+ * @author Maxime Steinhausser <maxime.steinhausser@gmail.com>
+ *
+ * @internal
+ */
+final class DelayedDumper implements DataDumperInterface, EventSubscriberInterface
+{
+    private $charset;
+    private $flags;
+    private $displayOptions;
+
+    /** @var Data[] */
+    private $data = array();
+
+    /** @var CliDumper */
+    private $cliDumper;
+
+    public function __construct(CliDumper $cliDumper, string $charset = null, int $flags = 0, array $displayOptions = array())
+    {
+        $this->charset = $charset;
+        $this->flags = $flags;
+        $this->displayOptions = $displayOptions;
+        $this->cliDumper = $cliDumper;
+    }
+
+    public function dump(Data $data)
+    {
+        // only delay web requests
+        if (!\in_array(PHP_SAPI, array('cli', 'phpdbg'), true)) {
+            $this->data[] = $data;
+
+            return;
+        }
+
+        $this->cliDumper->dump($data);
+    }
+
+    public function __destruct()
+    {
+        if ($this->data) {
+            $h = headers_list();
+            $i = count($h);
+            array_unshift($h, 'Content-Type: '.ini_get('default_mimetype'));
+            while (0 !== stripos($h[$i], 'Content-Type:')) {
+                --$i;
+            }
+
+            if (!\in_array(PHP_SAPI, array('cli', 'phpdbg'), true) && stripos($h[$i], 'html')) {
+                $dumper = new HtmlDumper('php://output', $this->charset);
+                $dumper->setDisplayOptions($this->displayOptions);
+            } else {
+                $dumper = new CliDumper('php://output', $this->charset, $this->flags);
+            }
+
+            foreach ($this->data as $data) {
+                $dumper->dump($data);
+            }
+
+            $this->data = array();
+        }
+    }
+
+    public function onKernelResponse(FilterResponseEvent $event)
+    {
+        if (!$event->isMasterRequest()) {
+            return;
+        }
+
+        $response = $event->getResponse();
+        $request = $event->getRequest();
+
+        // In all conditions that remove the web debug toolbar, dumps must be written on the output.
```

Comment:
why would this service check the debug toolbar being injected or no ? This class is not related to the data collector AFAICT

Reply:
You're right it's not directly related to the data collector, but this whole patch aims to ensure the exact same behavior exists whether the server dumper is wired or not.
When not wired and the toolbar being injected, dumps are already rendered in the toolbar and considered not to be output twice by the collector.